python/morphology.cpp

```diff
@@ -13,6 +13,102 @@
 
 namespace pyarb {
 
+arb::segment_tree load_swc_allen(const std::string& fname, bool no_gaps=false) {
+        std::ifstream fid{fname};
+        if (!fid.good()) {
+            throw pyarb_error(util::pprintf("can't open file '{}'", fname));
+        }
+        try {
+            using namespace arb;
+            auto records = parse_swc_file(fid);
+
+            // Assert that the file contains at least one sample.
+            if (records.empty()) {
+                throw pyarb_error("Allen SWC: empty file");
+            }
+
+            // assert that root sample has tag 1.
+            if (records[0].tag!=1) {
+                throw pyarb_error("Allen SWC: the soma record does not have tag 1");
+            }
+
+            // Assert that all non-root samples have a tag of 2, 3, or 4.
+            auto it = std::find_if(
+                        records.begin()+1, records.end(),
+                        [](auto& r){return r.tag<2 || r.tag>4;});
+            if (it!=records.end()) {
+                throw pyarb_error(
+                        "Allen SWC: every record must have a tag of 2, 3 or 4, except for the first which must have tag 1");
+            }
+
+            // Assert that all samples have the same tag as their parent, except
+            // those attached to the soma.
+            it = std::find_if(
+                        records.begin()+1, records.end(),
+                        [&records](auto& r){auto p = r.parent_id; return p && r.tag!=records[p].tag;});
+            if (it!=records.end()) {
+                throw pyarb_error(
+                        "Allen SWC: every record not attached to the soma must have the same tag as its parent");
+            }
+
+            // Translate the morphology so that the soma is centered at the origin (0,0,0)
+            mpoint sloc{records[0].x, records[0].y, records[0].z, records[0].r};
+            for (auto& r: records) {
+                r.x -= sloc.x;
+                r.y -= sloc.y;
+                r.z -= sloc.z;
+            }
+
+            segment_tree tree;
+
+            // Model the spherical soma as a cylinder with length=2*radius.
+            // The cylinder is centred on the origin, and extended along the y axis.
+            double soma_rad = sloc.radius;
+            tree.append(mnpos, {0, -soma_rad, 0, soma_rad}, {0, soma_rad, 0, soma_rad}, 1);
+
+            // Build branches off soma.
+            std::unordered_map<msize_t, msize_t> pmap;
+            std::set<msize_t> unused_samples;
+            const auto nrec = records.size();
+            for (unsigned i=1; i<nrec; ++i) {
```

Comment:
That will fall over if size == 1, right? Maybe just return just before the loop here?

Reply:
This isn't a problem. I add the soma associated with sample 0 above, and this loop will not execute.

Reply:
Right, that's true.